[DOCS-11836] How Step Duration is Determined in Browser Tests (#31603)
DataDog/documentation: content/en/synthetics/guide/step-duration.md

- Delays may occur when synchronous or deferred scripts prevent DOM completion.

**`loadEventEnd`**: Confirms the browser has finished firing the `load` event
- Occurs after all static resources (HTML, CSS, JavaScript, images) finish loading.
- Additional requests (AJAX calls, lazy loading, ads, analytics) may continue after this event.

These two signals work together to verify the page is fully rendered before the Synthetics Worker proceeds.

### Condition 2: Pending network requests

The Synthetics Worker also monitors pending network requests to determine step completion:

- Steps cannot complete while network requests are still active.
- A 10-second timeout prevents indefinite waiting.
- Certain requests are ignored (favicon pings, tracking URLs) to avoid unnecessary delays.

## Troubleshooting

### LCP impact on fully loaded time

Largest Contentful Paint (LCP) indirectly impacts step completion. While the Synthetics Worker doesn't explicitly wait for LCP, steps requiring interaction with LCP-related elements (such as buttons, large images, or text blocks) cannot proceed until those elements are fully rendered and ready.

### Deeper insights into step duration

> Test results may not always reveal why steps take longer than expected. For comprehensive analysis:

Author: ack, have added

---

> title: How Step Duration is Determined in Browser Tests

Reviewer: ```suggestion
title: How Step Duration is Determined in Browser Tests
description: Learn how Synthetic Monitoring determines when a browser test step is fully loaded and what factors influence step duration timing.
```

further_reading:
- link: "/synthetics/browser_tests"
  tag: "Documentation"
  text: "Learn more about Synthetic Monitoring Browser Testing"
- link: "/synthetics/guide/rum-to-synthetics/"
  tag: "Documentation"
  text: "Generate Synthetic Browser Tests From Session Replay"

---

## Overview

> Step duration in Synthetic Monitoring is the time a Browser Test step takes to be considered fully loaded. In most cases, a step represents a user action (for example, click and type). The Synthetics worker emulates the action, then waits for the page to finish loading and network requests to complete.

Reviewer: ```suggestion
Step duration in Synthetic Monitoring is the time a Browser Test step takes to be considered fully loaded. In most cases, a step represents a user action (such as click and type). The Synthetics worker emulates the action, then waits for the page to finish loading and network requests to complete.
```

Step duration includes:
- The action itself
- Page loading time  
- Network request completion time

Step duration does **not** include overhead such as finding the step element, taking screenshots, or creating snapshots on errors.

A page is considered fully loaded when:
- The [frame](#understanding-frames) has finished loading, _and_
- There are no requests still in flight.

In Synthetic Monitoring Browser Tests, step duration and fully loaded time represent the same concept:

{{< img src="synthetics/guide/step-duration/step_duration_2.png" alt="Side panel of a Synthetic Monitoring step highlighting fully loaded and step duration times" style="width:100%" >}}

## Understanding frames

A frame is a browser section that renders a web page. Each page contains one main frame (the primary [browsing][1] context) and may include child [frames][2] for content like advertisements, embedded widgets, or cross-origin elements.

> - **Main frame** = The primary image in a picture frame (your webpage).

Reviewer: ```suggestion
- **Main frame**: The primary image in a picture frame (your webpage).
```

> - **Child frames** = Smaller images within the main picture (iframes).

Reviewer: ```suggestion
- **Child frames**: Smaller images within the main picture (iframes).
```
